src/pages/repository/content.js

```diff
@@ -28,9 +51,95 @@ class Content extends React.Component {
     }
   }
 
+  componentDidUpdate() {
+    const {builds, state} = this.props;
+
+    // populating build filter tag suggestions
+    if (builds && (state == undefined || state.suggestions == undefined || state.suggestions.length == 0)) {
+      this.setFilterSuggestions();
+    }
+  }
+
+  setFilterSuggestions() {
+    const {owner, name} = this.props.params;
+    const {builds} = this.props;
+
+    var dict = {};
+    var item;
+
+    Object.keys(builds).forEach(k => {
+      item = builds[k];
+      dict['author:' + item.author] = true;
+      dict['branch:' + item.branch] = true;
+      dict['status:' + item.status] = true;
+      dict['event:' + item.event] = true;
+      if(item.ref.includes('tags')) {
+        dict['tag:' + item.ref.replace('refs/tags/', '')] = true;
+      }
+      if(item.deploy_to !== '') {
+        dict['deploy_to:' + item.deploy_to] = true;
+      }
+    });
+
+    var suggestions = Object.keys(dict).sort((a, b) => {return a > b;});
+    suggestions = suggestions.map(item => {return {label:item, value:item};});
+    events.emit(BUILD_FILTER_SUGGESTIONS, {owner, name, suggestions});
+  }
+
+  filterCriteria() {
+    const {state} = this.props;
+    const tags = ['tag', 'branch', 'author', 'status', 'event', 'deploy_to'];
+
+    var predicates = {};
+
+    var filters = state.filter.split(';');
+    filters.forEach(tag => {
+      var parts = tag.split(':');
+      if(!tags.includes(parts[0])) { 
+        return; 
+      }
+
+      var p;
+      switch(parts[0]) {
+      case 'branch':
+      case 'author':
+      case 'event':
+      case 'status':
+      case 'deploy_to':
+        p = 'o.'+parts[0]+'.includes("'+parts[1]+'")';
+        break;
+      case 'tag':
+        p = 'o.ref.includes("refs/tags/'+parts[1]+'")';
+        break;
+      }
+
+      if(predicates.hasOwnProperty(parts[0])) {
+        predicates[parts[0]].push(p);
+      } else {
+        predicates[parts[0]] = [p];
+      }
+    });
+
+    var func_body;
+    if(Object.keys(predicates).length > 0) {
+      var conditions = [];
+      Object.keys(predicates).forEach(k => {
+        conditions.push('(' + predicates[k].join(' || ') + ')');
+      });
+      func_body = 'return ' + conditions.join(' && ') + ';';
+    }
+
+    return new Function('o', func_body);
```

Comment:
Yes this is really bad. You are basically constructing JS code strings and `eval` it for every single suggestion cycle...